compiler/modules/types/list.bal

```diff
@@ -572,11 +560,56 @@ function listSubtypeIntersect(Context cx, SubtypeData t1, SubtypeData t2) return
 }
 
 function listSubtypeDiff(Context cx, SubtypeData t1, SubtypeData t2) returns SubtypeData {
-    Bdd b = memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
-    if b != false && cx.listMemo[b] == () {
-        //io:println("fresh BDD: ", bddToString(b));
+    MemoBddCache cache = new (cx.listMemo);
+    // This is checking if t1 and t2 are disjoint (i.e intersection(t1, t2) is empty). If so diff(t1, t2) == t1.
+    if listIsEmptySimple(cx, bddIntersect(cache, <Bdd> t1, <Bdd> t2)) == true {
+        return t1;
+    }
+    return memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
+}
+
+function listIsEmptySimple(Context cx, Bdd bdd) returns boolean? {
+    BddMemo? m = cx.listMemo[bdd];
+    if m != () && m.empty is boolean {
+        return <boolean>m.empty;
+    }
+    boolean empty = bddEvery(cx, bdd, (), (), listFormulaIsDefinitelyEmpty);
+    if empty == true {
+        cx.listMemo.put({ bdd, empty }); // handle both memoized values being () and not in memo
+        return true;
+    }
+    return ();
+}
+
+function listFormulaIsDefinitelyEmpty(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
+    if neg != () || pos == () {
+        return false;
+    }
+    var { members, rest } = cx.listAtomType(pos.atom);
+    Conjunction? current = pos.next;
+    while current != () {
+        var { members: currentMembers, rest: currentRest } = cx.listAtomType(current.atom);
+        var intersection = listIntersectWith(cx, members, rest, currentMembers, currentRest); 
+        if intersection == () {
+            return false;
```

Comment:
Shouldn't this be `true`. listIntersectWith returns nil when the result is empty because the list lengths are inconsistent.

Reply:
fixed